### ISSUE TITLE
Rewrite verification without exceptions

### ISSUE THREAD

**Issue by webmiche** [xdslproject/xdsl#79]

**Comment by webmiche** [comment]
I think Lawrence (@wence-) had some opinions on this. Maybe he can repeat them here/share ressources, that explain them.

**Comment by math-fehr** [comment]
So Lawrence point was that a function with a signature `(Operation) -> None` looks like something that could be removed.
I kind of agree with this, though in terms of a usability perspective, in a jupyter notebook, having such operation is really nice, compared to something like: `verif_result = op.verify(); if verif_result is error: raise VerifException(verif_result)`
So for now, I'm not really sure how to improve this, besides renaming `verify` to something like `assert_invariants`, which is more explicit

**Comment by wence-** [comment]
In a Haskell-like world, assertion has type `Data -> IO ()`, whereas "verification/parsing" has type `Data -> IO StructuredData`. In Python, the pattern I would use is:
```
def verify(self):
    if some_condition_does_not_hold(self):
        raise VerificationError(...)
    return more_structured_data(self)
```

And use like:

```
op = op.verify()
```

For more explanation/background see, for example https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/ and/or http://cliffle.com/blog/rust-typestate/

**Comment by math-fehr** [comment]
So here, in the specific context of MLIR, we can't really use the typestate pattern from my understanding.
The reason is that while after a `verify` the data is structured, during a transformation the data is not always structured, and that depends on the transformation itself.
We don't want users to check every time if the data is structured or not, especially since there is multiple levels of structure. For instance, you could have the right number of operands/results, but the wrong types on the arguments.
So in our cases, we would like an assertion rather than a verification. Though since we are in Python, I'm not completely sure if we can improve this. We could try to return an error instead of throwing one, but the user experience is I think way better if they can just write `op.assert_invariants()`.

**Comment by wence-** [comment]
I think you can just signal intent in that case by letting verify return self. That way the assignment is kind of a no-op, but it can make the type checker complain (and opens the door to returning something other than self later).

**Comment by math-fehr** [comment]
Hmm, the problem I would see with this approach is that if we do something like this, it has to be an in-place change, not a pure function.
This is because the IR has multiple cycles of references, so if we happen to modify an operation we need to update every reference.
Which is why I believe that just changing the name from `verify` to `assert_invariants` would be better, since we communicate then that this is just an assert.

**Comment by wence-** [comment]
Ah ok. I don't quite understand the details of the implementation, I think the name change is probably good. As an implementation detail, prefer to raise AssertionError over plain assert (the latter is removed when running the interpreter with -O)

**Comment by math-fehr** [comment]
Yes, I'm working at the same time to improve error reporting, so we will have a custom error for those cases!
Such that we have either an inline error in the IR, or that we print the line where the operation is failing.
Also, we may want to catch the error
